File: LinkedListMiddle.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        //two pointers - one twice as fast
        ListNode pointerOne = head;
        ListNode pointerTwo = head;
        
        
        //while pointer two.next is not null -> 
        //if pointer two.next.next is null -> return pointer one.next
        //else move pointerTwo two spots forward and pointerOne one spot forward
        while(pointerTwo.next != null) {
            if(pointerTwo.next.next == null) return pointerOne.next;
            pointerTwo = pointerTwo.next.next;
            pointerOne = pointerOne.next;
        }
        
        //return pointerOne
        return pointerOne;
        
        
    }
}

File: LinkedListMiddle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def middleNode(self, head):
        pointer1 = head
        pointer2 = head
        while not pointer2.next is None:
            if pointer2.next.next is None: return pointer1.next
            else:
                pointer1 = pointer1.next
                pointer2 = pointer2.next.next
        return pointer1


